src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -0,0 +1,459 @@
+                + backPart
+                + ANSI.noBackground()
+                + " " + clock + " "
+                + String.format("%d:%02d:%02d.%03d", duration.toHoursPart(), duration.toMinutes(), duration.toSecondsPart(), duration.toMillisPart())
+                + " "
+                ;
+
+            writer.println(line);
+        }
+
+        private String threadLabel() {
+            if (threadName.isEmpty()) {
+                return "";
+            }
+            else if ("main".equals(threadName)) {
+                return "";
+            }
+            
+            return threadName + ": ";
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            return obj instanceof ProgressBar 
+                && ((ProgressBar) obj).name.equals(name)
+                && ((ProgressBar) obj).threadId == threadId
+                ;
+        }
+
+        @Override
+        public int hashCode() {
+            return name.hashCode() + 31 * (int) threadId;
+        }
+
+        public void done() {
+            this.current = Math.min(current, max);
+            this.duration = Duration.between(startTime, Instant.now());
+            this.message = name;
+        }
+    }
+
+    /**
+     * Clean the screen, ready for the next output (normal output or the next progress bars or both).
+     * We use fancy ANSI codes here to move the cursor and clean the screen to remove previous versions
+     * of the bars
+     */
+    private void eraseBars() {
+        if (bars.isEmpty()) {
+            // writer.println();
+            return;
+        }
+            
+        writer.write(ANSI.hideCursor());
+        writer.write(ANSI.moveUp(bars.size())); 
+        writer.write(ANSI.clearToEndOfScreen()); 
+        writer.write(ANSI.showCursor());
+        writer.flush();
+    }
+
+    /**
+     * ANSI escape codes convenience functions
+     */
+    private static class ANSI {
+        static String moveUp(int n) {
+            return "\u001B[" + n + "F";
+        }
+
+        public static String darkBackground() {
+            return "\u001B[48;5;242m";
+        }
+
+        public static String noBackground() {
+            return "\u001B[49m";
+        }
+
+        public static String lightBackground() {
+            return "\u001B[48;5;249m";
+        }
+
+        static String moveDown(int n) {
+            return "\u001B[" + n + "E";
+        }
+
+        static String clearToEndOfScreen() {
+            return "\u001B[0J";
+        }
+
+        static String hideCursor() {
+            return "\u001B[?25l";
+        }
+
+        static String showCursor() {
+            return "\u001B[?25h";
+        }
+    }
+
+    /**
+     * Simply print the bars. No cursor movement here. Hiding the cursor prevents flickering.
+     */
+    private void printBars() {
+        writer.write(ANSI.hideCursor());
+        for (var pb : bars) {
+            pb.write();
+        }
+        writer.write(ANSI.showCursor());
```

Comment:
I would prefer it if every time we hide the cursor, we also show the cursor in a finally block. (in other places as well)